research: compile-time verification attribute macros

research: compile-time verification attribute macros

## Problem

Even with a structural checker, someone can add a new `pub fn` and forget to add a proof. The checker catches it in CI — but that's a 10-minute feedback loop. We want compile-time enforcement where possible.

## Proposed solution

A lightweight proc-macro crate (`logfwd-lint-attrs` or extending the planned one) that provides zero-cost attributes for verification linkage:

```rust
#[verified(kani = "verify_encode_varint_contract")]
pub fn encode_varint(buf: &mut Vec<u8>, value: u64) { ... }

#[trust_boundary]
pub fn decode_protobuf_to_batch(body: &[u8]) -> Result<RecordBatch, InputError> { ... }

#[state_machine]
pub struct PipelineMachine<S, C> { ... }
```

These attributes:
1. Compile to zero-cost doc markers (`#[doc = "__logfwd_verified_kani_verify_encode_varint_contract__"]`)
2. Are readable by the `xtask verify` tool
3. For `#[verified(kani = "name")]`: generate a `#[cfg(kani)] const _: () = { ... }` that references the proof function — **compile error if proof doesn't exist**

## Investigation scope

### 1. Design the attribute set
- What attributes are needed? `#[verified]`, `#[trust_boundary]`, `#[state_machine]`, `#[allow(logfwd::unproven)]`
- What arguments does each take?
- How do they compose with existing attributes (`#[cfg(kani)]`, `#[kani::proof]`)?

### 2. Implement the compile-time linkage
The key mechanism: `#[verified(kani = "verify_foo")]` generates:
```rust
#[cfg(kani)]
const _VERIFY_LINKAGE: () = {
    // This line fails to compile if verify_foo doesn't exist
    let _ = crate::module::verification::verify_foo as fn();
};
```

Investigate:
- Does this work across module boundaries?
- Does it work with `#[cfg(kani)]` gating?
- What happens when the proof is in a different crate?
- Can we make the error message clear? ("function `encode_varint` requires Kani proof `verify_foo` but it was not found")

### 3. Design the `#[trust_boundary]` attribute
- Should this generate a marker that the fuzz target manifest checker reads?
- Or should it generate a compile-time check that a fuzz target function exists?
- How do we reference fuzz targets (which are in a separate crate)?

### 4. Assess the annotation burden
- How many functions need `#[verified]`? (count pub fns in logfwd-core)
- Is it better to annotate functions that ARE proven, or exempt functions that AREN'T?
- For logfwd-core where "all pub fn need proofs" is the rule, should there be a crate-level attribute instead?

### 5. Prototype and test
- Build the proc-macro crate
- Annotate 5-10 functions in logfwd-core
- Verify compile-time error when proof is missing
- Verify zero runtime cost
- Check interaction with cargo kani, cargo test, cargo build

## What the investigation should produce

- Attribute macro crate implementation
- Compile-time linkage mechanism with clear error messages
- Assessment of annotation burden and opt-in vs opt-out strategy
- Integration plan with xtask verify
- 5-10 function annotations as proof of concept

## Files to read

- The planned `crates/logfwd-lint-attrs/` structure (from dylint research)
- `crates/logfwd-core/src/otlp.rs` — example functions to annotate
- `crates/logfwd-core/src/cri.rs` — example functions to annotate

## Related issues

- #1786, #1787, #1788 Kani contract strategy

## Research Results: Compile-Time Verification Attribute Macros

### TL;DR

- **Prototype works.** `#[verified(kani = "proof_fn")]` produces a compile error under `cfg(kani)` when the proof function is missing, with a clear error message pointing at the annotation site.
- **Zero runtime cost confirmed.** Doc markers (`#[doc = "__logfwd_verified_kani_..."]`) have no codegen effect.
- **67 functions** in logfwd-core would need annotation. Recommend **opt-in** (`#[verified]`) rather than opt-out (`#[allow_unproven]`).
- **Kani's own contract system** (`#[kani::requires]`/`#[kani::ensures]`/`#[kani::proof_for_contract]`) is complementary, not a replacement. Contracts specify *what* a function guarantees; `#[verified]` ensures *that* a proof exists.
- **Crate-level `#![require_proofs]`** is not feasible with proc macros alone (can't see other items in the crate). Would require a dylint or build script.

### 1. Compile-Time Linkage Mechanism

#### How it works

The `#[verified(kani = "verify_foo")]` proc macro generates:

```rust
// Zero-cost doc marker (survives to HIR, readable by dylint/xtask)
#[doc = "__logfwd_verified_kani_verify_foo__"]

// Compile-time existence check (only under cargo kani)
#[cfg(kani)]
const _VERIFIED_LINKAGE_FOO: () = {
    let _ = self::verification::verify_foo as *const ();
};

// Original function, unchanged
pub fn foo(x: u64) -> u64 { x + 1 }
```

#### Error messages

**Missing proof function** (verification module exists but proof absent):
```
error[E0425]: cannot find value `verify_missing_proof` in module `self::verification`
 --> src/lib.rs:8:1
  |
8 | #[verified(kani = "verify_missing_proof")]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `self::verification`
```

**Missing verification module entirely:**
```
error[E0433]: could not find `verification` in the crate root
 --> src/lib.rs:5:1
  |
5 | #[verified(kani = "verify_foo")]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ could not find `verification` in the crate root
```

Both errors point directly at the `#[verified]` annotation line. The developer immediately sees which function is missing a proof and what the proof should be named.

#### Cross-module behavior

The const assertion uses `self::verification::proof_fn`, which resolves to the `verification` submodule of the *current* module. This matches the existing convention in logfwd-core where proofs live in `#[cfg(kani)] mod verification { ... }` at the bottom of each file.

For functions in `otlp.rs`, the proof must be in `otlp.rs`'s verification module. This is correct -- proofs colocated with the code they verify.

Cross-crate proof references (e.g., proofs for logfwd-core functions living in logfwd-arrow) are **not supported** by this mechanism. This is intentional -- Kani proofs should live next to the functions they verify.

#### When it fires

The `#[cfg(kani)]` gate means the linkage check only runs during `cargo kani`. Normal `cargo build`, `cargo test`, and `cargo clippy` are unaffected. This is the right behavior: you don't want normal builds to fail because Kani isn't installed.

However, this means the check has the same feedback latency as CI (wherever `cargo kani` runs). To get *true* compile-time enforcement during local development, we'd need to drop the `#[cfg(kani)]` gate and reference a stub instead. See "Alternative: always-on linkage" below.

### 2. Doc Marker Pattern

The doc marker approach is identical to the existing `#[hot_path]` pattern in `logfwd-lint-attrs`:

- Proc macro prepends `#[doc = "__logfwd_verified_kani_PROOF_NAME__"]`
- The doc attribute is consumed by rustdoc but has zero codegen effect
- A dylint lint (like `hot_path_no_alloc`) can read the marker via `cx.tcx.hir_attrs(hir_id)` and `attr.value_str()`
- An xtask tool can parse source with `syn` and read the `#[verified(...)]` attribute directly (before expansion)

Both discovery paths work. The xtask path is simpler (no dylint toolchain needed) and sufficient for the "does every pub fn have a proof?" check.

### 3. Annotation Burden Assessment

#### Function census

| File | pub fn | pub(crate) fn | Total |
|------|--------|---------------|-------|
| otlp.rs | 15 | 0 | 15 |
| checkpoint_tracker.rs | 10 | 0 | 10 |
| reassembler.rs | 6 | 4 | 10 |
| structural.rs | 8 | 0 | 8 |
| structural_iter.rs | 6 | 0 | 6 |
| framer.rs | 5 | 0 | 5 |
| cri.rs | 3 | 2 | 5 |
| scan_config.rs | 4 | 0 | 4 |
| byte_search.rs | 3 | 0 | 3 |
| json_scanner.rs | 1 | 0 | 1 |
| **Total** | **61** | **6** | **67** |

#### Existing proof coverage

There are ~96 Kani proof harnesses across logfwd-core. Many functions have multiple proofs (e.g., `encode_varint` has `verify_varint_len_matches_encode`, `verify_varint_format_and_roundtrip`, `verify_varint_no_panic`). The `#[verified]` attribute would reference the primary proof, with secondary proofs being implicitly linked.

#### Opt-in vs opt-out

**Recommendation: opt-in (`#[verified]` on proven functions).**

Reasons:
- **67 annotations is manageable.** A single pass through the codebase adds them all.
- **Self-documenting.** Reading `#[verified(kani = "verify_encode_varint")]` immediately tells you which proof covers this function.
- **No false negatives.** An opt-out system (`#[allow_unproven]`) risks someone forgetting both the proof AND the exemption.
- **Gradual adoption.** Can annotate existing proven functions first, then make the xtask enforce "all pub fn must be `#[verified]` or `#[allow_unproven]`" later.

For functions genuinely exempt from proofs (trivial accessors, trait impls proven transitively, async code using proptest), use `#[allow_unproven(reason = "...")]`.

#### Crate-level `#![require_proofs]`

Not feasible with proc macros alone. A proc macro operates on one item at a time and cannot see the full crate. Options:

1. **Dylint lint** -- can iterate all `fn` items in the crate, check for `#[verified]`/`#[allow_unproven]`, and flag uncovered functions. This is the most powerful approach but adds the dylint toolchain dependency.
2. **xtask script** -- parse source with `syn`, find all `pub fn`, check for attributes. The existing `verify_proof_coverage.py` script does something similar but with regex; a Rust xtask with syn would be more robust.
3. **Build script** -- a `build.rs` that scans `src/*.rs` and fails if uncovered functions are found. Fragile (regex-based) but zero extra deps.

**Recommendation:** Start with xtask (option 2), graduate to dylint (option 1) if the xtask becomes insufficient.

### 4. Relationship to Kani Contracts

Kani's contract system (`#[kani::requires]`, `#[kani::ensures]`, `#[kani::proof_for_contract]`) is **complementary** to `#[verified]`:

| Aspect | `#[verified]` | Kani contracts |
|--------|---------------|----------------|
| Purpose | Ensure a proof *exists* | Specify what the proof *checks* |
| Granularity | "has a proof" | Pre/postconditions |
| Compile-time check | Proof function exists | Contract is satisfiable |
| Runtime cost | Zero | Zero (cfg(kani) gated) |
| Status | Custom proc macro | Experimental in Kani |

logfwd-core already uses Kani contracts on 3 functions (`compute_real_quotes`, `parse_4digits`, `parse_2digits`, `days_from_civil`) with `#[kani::ensures]` and `#[kani::proof_for_contract]`. These provide stronger guarantees than `#[verified]` but require more effort to write.

**Could we extend Kani's attributes instead?** No -- Kani's attributes are handled by its own compiler plugin and are not extensible via external proc macros. We cannot add custom attributes to the `kani::` namespace. The `#[verified]` attribute complements Kani's own attributes.

**Can they coexist?** Yes. A function can have both:

```rust
#[verified(kani = "verify_days_from_civil_contract")]
#[cfg_attr(kani, kani::requires(y >= 0 && y <= 9999))]
#[cfg_attr(kani, kani::ensures(|result: &i64| ...))]
pub fn days_from_civil(y: u16, m: u8, d: u8) -> i64 { ... }

#[cfg(kani)]
mod verification {
    #[kani::proof_for_contract(days_from_civil)]
    fn verify_days_from_civil_contract() { ... }
}
```

### 5. Alternative: Always-On Linkage (Without cfg(kani))

The `#[cfg(kani)]` gate means the linkage check only fires during `cargo kani` runs. For truly instant feedback, we could generate the const *without* the cfg gate:

```rust
#[cfg(kani)]
const _VERIFIED_LINKAGE_FOO: () = {
    let _ = self::verification::verify_foo as *const ();
};
```

becomes:

```rust
// Always checked, even in normal builds
const _VERIFIED_LINKAGE_FOO: () = {
    // Only needs the SIGNATURE to exist, not the proof body
};
```

**Problem:** The `verification` module itself is `#[cfg(kani)]`, so removing the gate from the const would cause "module not found" errors in normal builds.

**Solution:** Split the convention -- put proof *stubs* (just signatures, no bodies) in a non-cfg-gated module, and the actual proof bodies in `#[cfg(kani)]`. But this doubles the boilerplate.

**Recommendation:** Keep the `#[cfg(kani)]` gate. The xtask checker provides the "every push" enforcement; the compile-time check catches mismatches during `cargo kani` development.

### 6. Prototype Location

Working prototype at `/tmp/verified-attr-prototype/`:
- `verified-attrs/src/lib.rs` -- the proc-macro crate with `#[verified]`, `#[trust_boundary]`, `#[state_machine]`, `#[allow_unproven]`
- `demo-crate/src/lib.rs` -- happy path: 2 verified functions, trust boundary, state machine, exemption, Kani proofs, passing tests
- `demo-broken/src/lib.rs` -- error case: missing proof function, confirmed compile error
- `demo-no-mod/src/lib.rs` -- error case: missing verification module, confirmed compile error

All tests pass. Both error cases produce clear, actionable compiler errors pointing at the `#[verified]` annotation.

### 7. Implementation Plan

1. **Add `#[verified]` and `#[allow_unproven]` to `logfwd-lint-attrs`** -- extends the existing proc-macro crate, no new dependencies (the existing crate has zero deps, and we need `syn`/`quote` for argument parsing).
2. **Annotate all 67 pub/pub(crate) fns in logfwd-core** -- single PR, mechanical change. Exempt ~15 trivial accessors/constructors with `#[allow_unproven]`.
3. **Extend `xtask verify`** -- parse `#[verified]` attributes with syn, cross-reference against `#[kani::proof]` functions, report coverage gaps.
4. **Add to CI** -- `xtask verify --strict` fails if any pub fn lacks `#[verified]` or `#[allow_unproven]`.
5. **Later:** migrate the Python `verify_proof_coverage.py` to Rust xtask, add `#[trust_boundary]` and `#[state_machine]` enforcement.

### 8. Open Questions

- **Multiple proofs per function:** Should `#[verified(kani = "verify_encode_varint")]` reference just the primary proof, or should it support `#[verified(kani = ["verify_a", "verify_b"])]` for functions with multiple proofs?
- **proptest coverage:** Should there be `#[verified(proptest = "test_name")]` for functions proven by proptest instead of Kani? (e.g., `scan_streaming` which is too complex for Kani)
- **Naming enforcement:** Should the xtask enforce the `verify_<fn_name>_<property>` naming convention for proof harnesses?

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -40,6 +40,7 @@ default-members = [
     "crates/logfwd-types",
     "crates/logfwd-diagnostics",
     "crates/logfwd-core",
+    "crates/logfwd-lint-attrs",
     "crates/logfwd-arrow",
     "crates/logfwd-io",
     "crates/logfwd-output",
@@ -109,7 +110,9 @@ wasm-bindgen = "0.2"
 js-sys = "0.3"
 varsubst = { version = "0.0.1", default-features = false }
 xxhash-rust = { version = "0.8.15", features = ["xxh32", "xxh64"] }
-rustls = { version = "0.23", default-features = false, features = ["aws_lc_rs"] }
+rustls = { version = "0.23", default-features = false, features = [
+    "aws_lc_rs",
+] }
 rcgen = "0.14"
 
 [workspace.lints.clippy]

--- a/crates/logfwd-core/Cargo.toml
+++ b/crates/logfwd-core/Cargo.toml
@@ -9,17 +9,18 @@ publish = false
 doctest = false
 
 [dependencies]
+logfwd-lint-attrs = { version = "0.1.0", path = "../logfwd-lint-attrs" }
 memchr = { version = "2", default-features = false }
 wide = { version = "1.3.0", default-features = false }
 
 [dev-dependencies]
-logfwd-arrow = { version = "0.1.0", path = "../logfwd-arrow" }
-logfwd-test-utils = { version = "0.1.0", path = "../logfwd-test-utils" }
 arrow = { workspace = true }
 arrow-json = { workspace = true }
 bytes = "1"
 chrono = { version = "0.4.44", default-features = false }
 criterion = { version = "0.7", features = ["html_reports"] }
+logfwd-arrow = { version = "0.1.0", path = "../logfwd-arrow" }
+logfwd-test-utils = { version = "0.1.0", path = "../logfwd-test-utils" }
 memmap2 = "0.9"
 proptest = "1"
 sonic-rs = "0.5"

--- a/crates/logfwd-core/src/otlp.rs
+++ b/crates/logfwd-core/src/otlp.rs
@@ -116,9 +116,11 @@ pub const KEY_VALUE_VALUE: u32 = 2;
 // --- Protobuf wire format helpers ---
 
 use alloc::vec::Vec;
+use logfwd_lint_attrs::{allow_unproven, trust_boundary, verified};
 
 /// Encode a varint into buf at offset, return new offset.
 #[inline(always)]
+#[verified(kani = "verify_varint_len_matches_encode")]
 pub fn encode_varint(buf: &mut Vec<u8>, mut value: u64) {
     loop {
         if value < 0x80 {
@@ -133,6 +135,7 @@ pub fn encode_varint(buf: &mut Vec<u8>, mut value: u64) {
 /// Compute encoded varint length without writing.
 #[inline(always)]
 #[allow(clippy::match_overlapping_arm)]
+#[verified(kani = "verify_varint_len_matches_encode")]
 pub const fn varint_len(value: u64) -> usize {
     match value {
         0..=0x7F => 1,
@@ -150,12 +153,14 @@ pub const fn varint_len(value: u64) -> usize {
 
 /// Write a protobuf tag (field_number + wire_type).
 #[inline(always)]
+#[verified(kani = "verify_encode_tag")]
 pub fn encode_tag(buf: &mut Vec<u8>, field_number: u32, wire_type: u8) {
     encode_varint(buf, ((field_number as u64) << 3) | wire_type as u64);
 }
 
 /// Write a fixed64 field (tag + 8 bytes little-endian).
 #[inline(always)]
+#[verified(kani = "verify_encode_fixed64")]
 pub fn encode_fixed64(buf: &mut Vec<u8>, field_number: u32, value: u64) {
     encode_tag(buf, field_number, 1); // wire type 1 = 64-bit
     buf.extend_from_slice(&value.to_le_bytes());
@@ -163,6 +168,7 @@ pub fn encode_fixed64(buf: &mut Vec<u8>, field_number: u32, value: u64) {
 
 /// Write a varint field (tag + varint value).
 #[inline(always)]
+#[verified(kani = "verify_encode_varint_field")]
 pub fn encode_varint_field(buf: &mut Vec<u8>, field_number: u32, value: u64) {
     encode_tag(buf, field_number, 0); // wire type 0 = varint
     encode_varint(buf, value);
@@ -170,6 +176,7 @@ pub fn encode_varint_field(buf: &mut Vec<u8>, field_number: u32, value: u64) {
 
 /// Write a length-delimited field (tag + length + bytes).
 #[inline(always)]
+#[verified(kani = "verify_encode_bytes_field_content")]
 pub fn encode_bytes_field(buf: &mut Vec<u8>, field_number: u32, data: &[u8]) {
     encode_tag(buf, field_number, 2); // wire type 2 = length-delimited
     encode_varint(buf, data.len() as u64);
@@ -178,6 +185,7 @@ pub fn encode_bytes_field(buf: &mut Vec<u8>, field_number: u32, data: &[u8]) {
 
 /// Compute the encoded size of a length-delimited field (without writing).
 #[inline(always)]
+#[verified(kani = "verify_bytes_field_size")]
 pub const fn bytes_field_size(field_number: u32, data_len: usize) -> usize {
     let tag_size = varint_len(((field_number as u64) << 3) | 2);
     let len_size = varint_len(data_len as u64);
@@ -187,6 +195,7 @@ pub const fn bytes_field_size(field_number: u32, data_len: usize) -> usize {
 /// Write a fixed32 field (tag + 4 bytes little-endian).
 /// Used for LogRecord field 8 (`flags`), wire type 5.
 #[inline(always)]
+#[allow_unproven]
 pub fn encode_fixed32(buf: &mut Vec<u8>, field_number: u32, value: u32) {
     encode_tag(buf, field_number, 5); // wire type 5 = 32-bit fixed
     buf.extend_from_slice(&value.to_le_bytes());
@@ -198,6 +207,8 @@ pub fn encode_fixed32(buf: &mut Vec<u8>, field_number: u32, value: u32) {
 ///
 /// Returns `(value, new_pos)` or an error string if the input is truncated
 /// or the varint exceeds 10 bytes.
+#[allow_unproven]
+#[trust_boundary]
 pub fn decode_varint(buf: &[u8], pos: usize) -> Result<(u64, usize), &'static str> {
     let mut value: u64 = 0;
     let mut shift: u32 = 0;
@@ -220,6 +231,8 @@ pub fn decode_varint(buf: &[u8], pos: usize) -> Result<(u64, usize), &'static st
 }
 
 /// Decode a protobuf tag into `(field_number, wire_type, new_pos)`.
+#[allow_unproven]
+#[trust_boundary]
 pub fn decode_tag(buf: &[u8], pos: usize) -> Result<(u32, u8, usize), &'static str> {
     let (tag, new_pos) = decode_varint(buf, pos)?;
     let field_number = (tag >> 3) as u32;
@@ -229,6 +242,8 @@ pub fn decode_tag(buf: &[u8], pos: usize) -> Result<(u32, u8, usize), &'static s
 
 /// Skip a protobuf field value based on its wire type. Returns the new
 /// position after the field value.
+#[allow_unproven]
+#[trust_boundary]
 pub fn skip_field(buf: &[u8], wire_type: u8, pos: usize) -> Result<usize, &'static str> {
     match wire_type {
         0 => {
@@ -274,6 +289,7 @@ pub fn skip_field(buf: &[u8], wire_type: u8, pos: usize) -> Result<usize, &'stat
 ///
 /// Designed for zero-allocation decoding of `trace_id` (32 hex chars → 16 bytes)
 /// and `span_id` (16 hex chars → 8 bytes) on the hot encoding path.
+#[verified(kani = "verify_hex_decode_roundtrip")]
 pub fn hex_decode(hex_bytes: &[u8], out: &mut [u8]) -> bool {
     if hex_bytes.len() != out.len() * 2 {
         return false;
@@ -345,6 +361,7 @@ pub enum Severity {
 
 /// Fast severity lookup from first byte + length. No string comparison needed.
 #[inline(always)]
+#[verified(kani = "verify_parse_severity_no_false_positives")]
 pub fn parse_severity(text: &[u8]) -> (Severity, &[u8]) {
     // Exact case-insensitive match against the 6 standard severity strings
     // plus common aliases used by syslog and application logs.
@@ -373,12 +390,14 @@ pub fn parse_severity(text: &[u8]) -> (Severity, &[u8]) {
 /// Case-insensitive variable-length comparison for ASCII letters.
 /// Used by Kani proofs to verify parse_severity only matches standard levels.
 #[cfg(kani)]
+#[allow_unproven]
 fn eq_ignore_case_match(a: &[u8], b: &[u8]) -> bool {
     a.len() == b.len() && a.iter().zip(b.iter()).all(|(x, y)| x | 0x20 == y | 0x20)
 }
 
 /// Case-insensitive 3-byte comparison.
 #[inline(always)]
+#[allow_unproven]
 fn eq_ignore_case_3(a: &[u8], b: &[u8]) -> bool {
     a[0] | 0x20 == b[0] | 0x20 && a[1] | 0x20 == b[1] | 0x20 && a[2] | 0x20 == b[2] | 0x20
 }
@@ -388,6 +407,7 @@ fn eq_ignore_case_3(a: &[u8], b: &[u8]) -> bool {
 /// collisions for non-letters (e.g., `@` |0x20 = `` ` ``). Safe here because
 /// the comparison targets ("INFO", "WARN") are all ASCII letters.
 #[inline(always)]
+#[verified(kani = "verify_eq_ignore_case_4_no_false_positives_info")]
 fn eq_ignore_case_4(a: &[u8], b: &[u8]) -> bool {
     a[0] | 0x20 == b[0] | 0x20
         && a[1] | 0x20 == b[1] | 0x20
@@ -397,6 +417,7 @@ fn eq_ignore_case_4(a: &[u8], b: &[u8]) -> bool {
 
 /// Case-insensitive 5-byte comparison.
 #[inline(always)]
+#[verified(kani = "verify_eq_ignore_case_5_no_false_positives_error")]
 fn eq_ignore_case_5(a: &[u8], b: &[u8]) -> bool {
     a[0] | 0x20 == b[0] | 0x20
         && a[1] | 0x20 == b[1] | 0x20
@@ -407,6 +428,7 @@ fn eq_ignore_case_5(a: &[u8], b: &[u8]) -> bool {
 
 /// Case-insensitive 6-byte comparison.
 #[inline(always)]
+#[allow_unproven]
 fn eq_ignore_case_6(a: &[u8], b: &[u8]) -> bool {
     a[0] | 0x20 == b[0] | 0x20
         && a[1] | 0x20 == b[1] | 0x20
@@ -418,6 +440,7 @@ fn eq_ignore_case_6(a: &[u8], b: &[u8]) -> bool {
 
 /// Case-insensitive 7-byte comparison.
 #[inline(always)]
+#[allow_unproven]
 fn eq_ignore_case_7(a: &[u8], b: &[u8]) -> bool {
     a[0] | 0x20 == b[0] | 0x20
         && a[1] | 0x20 == b[1] | 0x20
@@ -430,6 +453,7 @@ fn eq_ignore_case_7(a: &[u8], b: &[u8]) -> bool {
 
 /// Case-insensitive 8-byte comparison.
 #[inline(always)]
+#[allow_unproven]
 fn eq_ignore_case_8(a: &[u8], b: &[u8]) -> bool {
     a[0] | 0x20 == b[0] | 0x20
         && a[1] | 0x20 == b[1] | 0x20
@@ -460,6 +484,7 @@ fn eq_ignore_case_8(a: &[u8], b: &[u8]) -> bool {
 ///
 /// Fractional seconds beyond 9 digits (nanosecond precision) are
 /// truncated — this is intentional as OTLP uses nanoseconds.
+#[verified(kani = "verify_parse_timestamp_compositional")]
 pub fn parse_timestamp_nanos(ts: &[u8]) -> Option<u64> {
     if ts.len() < 19 {
         return None;
@@ -579,6 +604,7 @@ pub fn parse_timestamp_nanos(ts: &[u8]) -> Option<u64> {
 #[allow(dead_code)]
 #[inline(always)]
 #[cfg_attr(kani, kani::ensures(|result: &u16| *result <= 9999))]
+#[verified(kani = "verify_parse_4digits_contract")]
 fn parse_4digits(s: &[u8], off: usize) -> u16 {
     if off + 4 > s.len() {
         return 0;
@@ -595,6 +621,7 @@ fn parse_4digits(s: &[u8], off: usize) -> u16 {
 #[allow(dead_code)]
 #[inline(always)]
 #[cfg_attr(kani, kani::ensures(|result: &u8| *result <= 99))]
+#[verified(kani = "verify_parse_2digits_contract")]
 fn parse_2digits(s: &[u8], off: usize) -> u8 {
     if off + 2 > s.len() {
         return 0;
@@ -611,6 +638,7 @@ fn parse_2digits(s: &[u8], off: usize) -> u8 {
 /// avoids the double-check that `parse_4digits` + external `is_ascii_digit`
 /// calls would perform.
 #[inline(always)]
+#[allow_unproven]
 fn parse_4digits_checked(s: &[u8], off: usize) -> Option<u16> {
     let (a, b, c, d) = (
         s[off].wrapping_sub(b'0'),
@@ -628,6 +656,7 @@ fn parse_4digits_checked(s: &[u8], off: usize) -> Option<u16> {
 /// not a digit. Single-pass: one `wrapping_sub` + compare per digit instead of
 /// a `is_ascii_digit` check followed by a separate subtraction.
 #[inline(always)]
+#[allow_unproven]
 fn parse_2digits_checked(s: &[u8], off: usize) -> Option<u8> {
     let a = s[off].wrapping_sub(b'0');
     let b = s[off + 1].wrapping_sub(b'0');
@@ -639,12 +668,14 @@ fn parse_2digits_checked(s: &[u8], off: usize) -> Option<u8> {
 
 /// Returns `true` if `year` is a leap year (Gregorian calendar).
 #[inline(always)]
+#[allow_unproven]
 fn is_leap_year(year: i64) -> bool {
     year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)
 }
 
 /// Returns the number of days in the given month (1-12) for the given year.
 #[inline(always)]
+#[allow_unproven]
 fn days_in_month(year: i64, month: u32) -> u32 {
     match month {
         1 | 3 | 5 | 7 | 8 | 10 | 12 => 31,
@@ -673,6 +704,7 @@ fn days_in_month(year: i64, month: u32) -> u32 {
     // Required so `stub_verified(days_from_civil)` keeps nanos arithmetic within u64.
     && *result <= 213_400
 ))]
+#[verified(kani = "verify_days_from_civil_contract")]
 pub fn days_from_civil(year: i64, month: u32, day: u32) -> i64 {
     let y = if month <= 2 { year - 1 } else { year };
     let m = if month <= 2 {

--- a/crates/logfwd-lint-attrs/Cargo.toml
+++ b/crates/logfwd-lint-attrs/Cargo.toml
@@ -10,6 +10,9 @@ description = "Attribute macros that mark functions for semantic lints (e.g., #[
 proc-macro = true
 
 [dependencies]
+proc-macro2 = "1"
+quote = "1"
+syn = { version = "2", features = ["full", "parsing"] }
 
 [lints]
 workspace = true

--- a/crates/logfwd-lint-attrs/src/lib.rs
+++ b/crates/logfwd-lint-attrs/src/lib.rs
@@ -123,3 +123,159 @@ fn prepend_marker(marker: &str, item: TokenStream) -> TokenStream {
     out.extend(item);
     out
 }
+
+// ---------------------------------------------------------------------------
+// Compile-time verification linkage attributes
+// ---------------------------------------------------------------------------
+
+use quote::{format_ident, quote};
+use syn::{
+    ItemFn, LitStr, MetaNameValue, Token,
+    parse::{Parse, ParseStream},
+    parse_macro_input,
+    punctuated::Punctuated,
+};
+
+struct VerifiedArgs {
+    kani: Option<LitStr>,
+    proptest: Option<LitStr>,
+}
+
+impl Parse for VerifiedArgs {
+    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
+        let pairs = Punctuated::<MetaNameValue, Token![,]>::parse_terminated(input)?;
+        let mut kani = None;
+        let mut proptest = None;
+
+        for pair in pairs {
+            if pair.path.is_ident("kani") {
+                match pair.value {
+                    syn::Expr::Lit(syn::ExprLit {
+                        lit: syn::Lit::Str(value),
+                        ..
+                    }) => {
+                        kani = Some(value);
+                    }
+                    _ => {
+                        return Err(syn::Error::new_spanned(
+                            pair,
+                            "expected string literal: kani = \"verify_name\"",
+                        ));
+                    }
+                }
+            } else if pair.path.is_ident("proptest") {
+                match pair.value {
+                    syn::Expr::Lit(syn::ExprLit {
+                        lit: syn::Lit::Str(value),
+                        ..
+                    }) => {
+                        proptest = Some(value);
+                    }
+                    _ => {
+                        return Err(syn::Error::new_spanned(
+                            pair,
+                            "expected string literal: proptest = \"test_name\"",
+                        ));
+                    }
+                }
+            } else {
+                return Err(syn::Error::new_spanned(
+                    pair.path,
+                    "unsupported key, expected `kani` or `proptest`",
+                ));
+            }
+        }
+
+        if kani.is_none() && proptest.is_none() {
+            return Err(syn::Error::new(
+                input.span(),
+                "expected at least one linkage: kani = \"...\" or proptest = \"...\"",
+            ));
+        }
+
+        Ok(Self { kani, proptest })
+    }
+}
+
+/// Marks a function as proof-verified.
+///
+/// `#[verified(kani = "verify_function_name")]` emits a compile-time check
+/// under `cfg(kani)` that references `self::verification::<proof_name>`.
+/// If the proof function does not exist, compilation fails at the annotation site.
+///
+/// Also accepts `proptest = "test_name"` to emit a `#[cfg(test)]` doc-marker
+/// for xtask/structural tooling.
+#[proc_macro_attribute]
+pub fn verified(args: TokenStream, item: TokenStream) -> TokenStream {
+    let args = parse_macro_input!(args as VerifiedArgs);
+    let function = parse_macro_input!(item as ItemFn);
+    let function_name = function.sig.ident.clone();
+    let marker_ident = format_ident!("__verified_marker_{}", function_name);
+
+    let kani_assertion = if let Some(kani_name) = args.kani {
+        let kani_proof = format_ident!("{}", kani_name.value());
+        quote! {
+            #[cfg(kani)]
+            #[doc(hidden)]
+            #[allow(non_upper_case_globals)]
+            const #marker_ident: () = {
+                let _proof: fn() = self::verification::#kani_proof;
+            };
+        }
+    } else {
+        quote! {}
+    };
+
+    let proptest_marker = if let Some(test_name) = args.proptest {
+        let value = test_name.value();
+        quote! {
+            #[cfg(test)]
+            #[doc(hidden)]
+            const _: &str = concat!("proptest:", #value);
+        }
+    } else {
+        quote! {}
+    };
+
+    TokenStream::from(quote! {
+        #function
+        #kani_assertion
+        #proptest_marker
+    })
+}
+
+/// Marker attribute for functions that process untrusted input.
+///
+/// Emits a hidden const marker for xtask/lint tooling. No runtime effect.
+#[proc_macro_attribute]
+pub fn trust_boundary(_args: TokenStream, item: TokenStream) -> TokenStream {
+    let function = parse_macro_input!(item as ItemFn);
+    let function_name = function.sig.ident.clone();
+    let marker_ident = format_ident!("__trust_boundary_marker_{}", function_name);
+
+    TokenStream::from(quote! {
+        #function
+        #[doc(hidden)]
+        #[allow(non_upper_case_globals)]
+        const #marker_ident: &str = concat!("trust_boundary:", stringify!(#function_name));
+    })
+}
+
+/// Marker attribute for explicit exemptions from proof linkage checks.
+///
+/// Documents that a function intentionally lacks a Kani proof (e.g.,
+/// internal helpers, decode paths where proptest provides coverage).
+/// No runtime effect.
+#[proc_macro_attribute]
+pub fn allow_unproven(_args: TokenStream, item: TokenStream) -> TokenStream {
+    let function = parse_macro_input!(item as ItemFn);
+    let function_name = function.sig.ident.clone();
+    let marker_ident = format_ident!("__allow_unproven_marker_{}", function_name);
+
+    TokenStream::from(quote! {
+        #function
+        #[doc(hidden)]
+        #[allow(non_upper_case_globals)]
+        const #marker_ident: &str = concat!("allow_unproven:", stringify!(#function_name));
+    })
+}

